kernel: Add Kconfig option to disable LTO for kernel sources

GTLin08 opened the pull request:
Some SoCs require kernel code to be placed in RAM, which makes link-time optimization (LTO) unsuitable for these files.
Disabling LTO allows the affected code to be linked as separate objects and placed in specific memory regions.
Running kernel code from RAM can improve execution performance, especially for timing-critical routines or context switch paths.

Another commit is to enable `KERNEL_NO_LTO `when `SOC_IT8XXX2_KERNEL_IN_RAM `and `LTO `are enabled on the IT8XXX2 chip.

fabiobaltieri commented on soc/ite/ec/it8xxx2/kernel_no_lto.cmake:
```diff
@@ -0,0 +1,38 @@
+                            PROPERTIES COMPILE_FLAGS \"-fno-lto -g\")")
+
+# Append the injected code to the end of the file
+file(APPEND ${KERNEL_CMAKE} "${INJECT_CODE}")
```
uh, why is this needed? The whole append/remove thing feels a bit hacky to be honest, hope we can find a different solution, you just need the `-fno-lto -g` in the build command line for the files above right?

GTLin08 replied:
The reason is that setting `-fno-lto` elsewhere (e.g. board-level CMake) didn’t work likely because of build order issues. Kernel sources were still built with LTO, so I had to apply it directly here to make sure those files compile without LTO and can reside in RAM.

fabiobaltieri replied:
ok so this is basically touching the source file to somehow convince cmake to build it again with the new flag?

GTLin08 replied:
Yes, that's correct, it temporarily modifies the kernel CMake file to ensure those specific sources are rebuilt with `-fno-lto`. After the build configuration is complete, the injected changes are automatically cleaned up to restore the original file state.

GTLin08 committed:
kernel: Add Kconfig option to disable LTO for kernel sources

Some SoCs require kernel code to be placed in RAM, which makes
link-time optimization (LTO) unsuitable for these files.
Disabling LTO allows the affected code to be linked as separate
objects and placed in specific memory regions.

Running kernel code from RAM can improve execution performance,
especially for timing-critical routines or context switch paths.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
soc: it8xxx2: Select KERNEL_NO_LTO only when LTO is enabled

Select KERNEL_NO_LTO only when LTO is enabled. This ensures proper
handling when kernel code is placed in RAM.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 commented on kernel/kernel_no_lto.cmake:
```diff
@@ -0,0 +1,44 @@
+  if(src MATCHES "\\.c$" AND
+     NOT src MATCHES "errno\\.c$" AND
+     NOT src MATCHES "fatal\\.c$" AND
+     NOT src MATCHES "init\\.c$")
```
Agreed, I’ve added them to the default LTO_ALLOWLIST. Thanks!

GTLin08 commented:
@fabiobaltieri Please revisit, thanks!

fabiobaltieri reviewed (dismissed):
@tejlmand does this look like an acceptable approach?

keith-zephyr commented on kernel/Kconfig:
```diff
@@ -1138,6 +1138,28 @@ endif # BOOTARGS
+	help
+	  List of kernel source filenames that should retain LTO even when
+	  CONFIG_KERNEL_NO_LTO is enabled.
+	  Example: CONFIG_KERNEL_LTO_ALLOWLIST="init.c errno.c fatal.c"
```
Did you want to make these files the default setting?  Otherwise you could expand the help text to suggest this list as a good default as these features don't benefit from the performance gain of running in RAM.

GTLin08 replied:
If we set default "init.c errno.c fatal.c" for CONFIG_KERNEL_LTO_ALLOWLIST, users would still need to override the entire string when adding more files. Kconfig does not support appending to the default value, so it cannot be extended incrementally.

GTLin08 replied:
Sure, once this PR is merged, I’ll create a follow-up PR to enable.

GTLin08 commented on kernel/kernel_no_lto.cmake:
```diff
@@ -0,0 +1,42 @@
+
+# Apply -fno-lto to all C source files, except for some initialization files
+# (e.g. init.c, errno.c, fatal.c) and those that are less critical to be
+# placed in RAM. These files can be excluded from -fno-lto by using
```
Done.

GTLin08 commented on kernel/Kconfig:
```diff
@@ -1138,6 +1138,28 @@ endif # BOOTARGS
 
+config KERNEL_NO_LTO
+	bool "Build selected kernel core files without LTO"
+	depends on LTO
```
> This should depend on XIP as well?

I've added the dependency on XIP as suggested.


> Also, are you using Zephyr Code Relocation to place kernel in RAM?

I'm not using Zephyr's generic code relocation framework. We relocate the kernel-related libraries into RAM using a platform-specific RAM region defined in our SoC linker script, so disabling LTO ensures the kernel objects keep their sections and can be placed correctly in RAM.

GTLin08 commented:
@tejlmand, Any feedback would be greatly appreciated.

GTLin08 commented:
Rebase.

tejlmand reviewed (changes requested):
Disabling LTO on kernel sources should be on library basis, not this level of granularity.

GTLin08 commented on kernel/Kconfig:
```diff
@@ -1159,6 +1159,19 @@ endif # BOOTARGS
 endmenu
 
+config KERNEL_NO_LTO
+	bool "Build selected kernel core files without LTO"
```
Remove the prompt.

GTLin08 commented on kernel/kernel_no_lto.cmake:
```diff
@@ -0,0 +1,50 @@
+
+# Check if there is a permitted kernel file that can enable LTO
+if(COMMAND kernel_lto_allow_list)
+  kernel_lto_allow_list(ADDITIONAL_ALLOWLIST)
```
Instead of calling `kernel_lto_allow_list()` I switched to a Kconfig solution. This makes the allowlist explicit, avoids ordering issues, and allows each board to define its own list through Kconfig.

GTLin08 commented on kernel/kernel_no_lto.cmake:
```diff
@@ -0,0 +1,50 @@
+    endif()
+
+  endif()
+endforeach()
```
I tried following the suggested direction of disabling LTO at the library level rather than granularity. However, based on the experiments I ran, LTO still appears to be applied to the kernel even after attempting to disable it.

My test was:  `target_compile_options(kernel PRIVATE -fno-lto)`
However, after compilation, the kernel-related `.c` files still appear to be optimized with LTO.

To investigate further, I checked where -flto is injected in the Zephyr CMake files:
```
zephyr/cmake/compiler/gcc/compiler_flags.cmake:
  set_compiler_property(PROPERTY optimization_lto -flto=auto)
zephyr/cmake/linker/ld/linker_flags.cmake:
  set_property(TARGET linker PROPERTY lto_arguments -flto=auto -fno-ipa-sra -ffunction-sections -fdata-sections)
```
Since `-flto=auto` is added through compiler properties and linker arguments at a higher level, target-level flags do not seem to override these settings in my tests. So far, the only approach that reliably disables LTO for specific kernel files is applying -fno-lto directly via `set_source_files_properties()` on a per–source-file basis.

GTLin08 replied:
@tejlmand, Any feedback would be greatly appreciated.
